src/python_osw_validation/__init__.py

```diff
@@ -58,6 +93,83 @@ def validate(self) -> ValidationResult:
                     zip_handler.remove_extracted_files()
                     return ValidationResult(False, self.errors)
 
+            # Validate data integrity
+            OSW_dataset = {}
+            for file in validator.files:
+                file_path = os.path.join(file)
+                osw_file = file_path.split('.')[-2]
+                OSW_dataset[osw_file] = gpd.read_file(file_path)
+
+            # Are all id's unique in each file? No need to check uniqueness across files yet since we do not have a global OSW ID format yet
+            for osw_file in OSW_dataset:
+                is_valid, duplicates = self.unique_id(OSW_dataset[osw_file])
+                if not is_valid:
+                    zip_handler.remove_extracted_files()
+                    self.errors.append(f"Duplicate _id's found in {osw_file} : {duplicates}")
+                    return ValidationResult(False, self.errors)
+
+            # Create sets of node id's and foreign keys to be used in validation
+            if "nodes" in OSW_dataset:
+                node_ids = set(OSW_dataset['nodes']['_id'])
+            else:
+                node_ids = set()
+
+            if "edges" in OSW_dataset:
+                node_ids_edges_u = set(OSW_dataset['edges']['_u_id'])
+                node_ids_edges_v = set(OSW_dataset['edges']['_v_id'])
+            else:
+                node_ids_edges_u = set()
+                node_ids_edges_v = set()
+
+            if "zones" in OSW_dataset:
+                node_ids_zones_w = set([item for sublist in OSW_dataset['zones']['_w_id'] for item in sublist])
+            else:
+                node_ids_zones_w = set()
+
+            # Do all node references in _u_id exist in nodes?
+            unmatched = node_ids_edges_u - node_ids
+            is_valid = len(unmatched) == 0
+            if not is_valid:
+                zip_handler.remove_extracted_files()
+                self.errors.append(f"Foreign key constraints for edge start nodes failed, _u_id's of unmatched nodes: {unmatched}")
+                return ValidationResult(False, self.errors)
+
+            # Do all node references in _v_id exist in nodes?
+            unmatched = node_ids_edges_v - node_ids
+            is_valid = len(unmatched) == 0
+            if not is_valid:
+                zip_handler.remove_extracted_files()
+                self.errors.append(f"Foreign key constraints for edge end nodes failed, _v_id's of unmatched nodes: {unmatched}")
```

Comment:
Fixed the message but didn't do the second part since it might be easier to locate the issue if we are specific about whether the issue is with _u_id vs _v_id vs _w_id